# StandardScaler

## Introduction
This notebook provides comprehensive documentation on **StandardScaler**,
a data standardization technique used in machine learning pipelines.

It includes :
- A **description** of the StandardScaler
- The **formula** used
- An explanation of **how each method works**
- **Code examples**
- A **comparison with Scikit-learn**

### Description
- The StandardScaler is used to **standardize data** by bringing it to the same scale
- It calculates the **mean** and **standard deviation** of each column
- It transforms values so that each column has :
  - A mean of **0**
  - A standard deviation of **1**

### Formula
- **X' = (X - mean) / standard deviation**

### Methods
The StandardScaler is based on **6 methods** :

#### 1. fit()
- **Role** : Calculates and stores the mean and standard deviation
- **Input** : Data array X
- **Process** :
  - Calculates mean using `np.mean()`
  - Calculates standard deviation using `np.std()`
  - Stores them in `self.mean_` and `self.std_`
- **Output** : Returns the scaler object

#### 2. transform()
- **Role** : Standardizes the data using the Z-score formula
- **Input** : Data array X
- **Process** :
  - Verifies that `fit()` has been called
  - Applies formula **X' = (X - mean) / standard deviation**
- **Output** : Standardized data in the same format as input

#### 3. fit_transform()
- **Role** : Combines `fit()` and `transform()` in one step
- **Process** :
  - Calls `fit(X)` to calculate mean and standard deviation
  - Calls `transform(X)` to standardize
- **Output** : Standardized data


#### 4. inverse_transform()
- **Role** : Restores data to its original scale
- **Input** : Standardized data
- **Process** :
  - Verifies that `fit()` has been called
  - Applies inverse formula **(X * standard deviation) + mean**
- **Output** : Original data


#### 5. _convert_to_array()
- **Role** : Converts input data into a NumPy array
- **Input** : Data X
- **Process** :
  - If X is a DataFrame → stores column names and returns values
  - Otherwise → converts directly to NumPy array
- **Output** : NumPy array


#### 6. _convert_to_dataframe()
- **Role** : Restores data to its original format
- **Input** : Transformed NumPy array
- **Process** :
  - If original data was a DataFrame → recreates DataFrame with column names
  - Otherwise → returns NumPy array unchanged
- **Output** : Data in original format


In [6]:
import sys
sys.path.append(r'C:\Users\LENOVO\ifri_mini_ml_lib\ifri_mini_ml_lib\preprocessing\preparation\scaler')

from standard_scaler import StandardScaler

print("Import successful !")


Import successful !


In [7]:
import numpy as np
import pandas as pd
import time
from sklearn.preprocessing import StandardScaler as SklearnStandardScaler

### Loading the dataset
df = pd.read_csv('base3.csv')

### Selecting 3 numerical columns
data = df[['Age', 'EstimatedSalary', 'Purchased']]

print("Original dataset:")
print(data.head(10))

### our implementation
start = time.time()
our_scaler = StandardScaler()
our_result = our_scaler.fit_transform(data)
our_time = time.time() - start

### Scikit-learn
start = time.time()
sklearn_scaler = SklearnStandardScaler()
sklearn_result = sklearn_scaler.fit_transform(data)
sklearn_time = time.time() - start

print("our StandardScaler :")
print(our_result)
print(f"Time : {our_time:.6f} seconds")

print("\nScikit-learn StandardScaler :")
print(sklearn_result)
print(f"Time : {sklearn_time:.6f} seconds")

print("\nIdentical results :", np.allclose(our_result, sklearn_result))

Original dataset:
   Age  EstimatedSalary  Purchased
0   19            19000          0
1   35            20000          0
2   26            43000          0
3   27            57000          0
4   19            76000          0
5   27            58000          0
6   27            84000          0
7   32           150000          1
8   25            33000          0
9   35            65000          0
our StandardScaler :
          Age  EstimatedSalary  Purchased
0   -1.781797        -1.490046  -0.745936
1   -0.253587        -1.460681  -0.745936
2   -1.113206        -0.785290  -0.745936
3   -1.017692        -0.374182  -0.745936
4   -1.781797         0.183751  -0.745936
..        ...              ...        ...
395  0.797057        -0.844019   1.340598
396  1.274623        -1.372587   1.340598
397  1.179110        -1.460681   1.340598
398 -0.158074        -1.078938  -0.745936
399  1.083596        -0.990844   1.340598

[400 rows x 3 columns]
Time : 0.002559 seconds

Scikit-learn StandardSc

## Analysis of Results

### 1. Transformation Results
- Both scalers produce **identical results (True)**
- The data is well centered around 0 :
  - Age values range from **-1.78 to +1.27**
  - Salaries range from **-1.49 to +1.34**
  - Purchased column : **-0.75** (not purchased) | **+1.34** (purchased)

### 2. Performance Metrics
- Our StandardScaler : **0.002559 seconds**
- Scikit-learn : **0.012293 seconds**
- Our implementation is approximately **5 times faster**

### 3. MinMaxScaler vs StandardScaler

**MinMaxScaler :**
- Scales values between **0 and 1**
- Formula : **(X - min) / (max - min)**
- Very sensitive to outliers
- Maintains fixed bounds **[0, 1]**
- Use when **data bounds are known**

**StandardScaler :**
- Centers values around **0**
- Formula : **(X - mean) / standard deviation**
- Less sensitive to outliers
- No fixed bounds
- Use when **data follows a normal distribution**

### 4. Conclusion
- Our StandardScaler produces correct results
- It is faster than Scikit-learn on this 400-row dataset
- It is a reliable alternative for data standardization in ML pipelines
